File: java/testsuite/ApplicationCases/SceneEditorApplication/src/model/history/LightTransformState.java
package model.history;

import vsdk.toolkit.common.Entity;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.environment.light.Light;

/**
Position of a light (the placement the editor can change).
*/
public final class LightTransformState implements EntityTransformState
{
    private final Light light;
    private final Vector3Dd position;

    private LightTransformState(Light light)
    {
        this.light = light;
        this.position = light.getPosition();
    }

    /**
    @param light light whose placement is captured
    @return the current placement of the light
    */
    public static LightTransformState capture(Light light)
    {
        return new LightTransformState(light);
    }

    @Override
    public Entity getEntity()
    {
        return light;
    }

    @Override
    public void restore()
    {
        light.setPosition(position);
    }

    @Override
    public boolean isSameState(EntityTransformState other)
    {
        if ( !(other instanceof LightTransformState state) ) {
            return false;
        }
        return state.light == light &&
            (position == null ? state.position == null : position.equals(state.position));
    }
}

File: cpp/testsuite/ApplicationCases/SceneEditorApplication/src/model/history/LightTransformState.cpp
#include "vsdk/toolkit/environment/light/Light.h"
#include "model/history/LightTransformState.h"

LightTransformState::LightTransformState(Light* light)
    : light(light), position(light->getPosition())
{
}

LightTransformState* LightTransformState::capture(Light* light)
{
    return new LightTransformState(light);
}

Entity* LightTransformState::getEntity() const
{
    return light;
}

void LightTransformState::restore() const
{
    light->setPosition(position);
}

bool LightTransformState::isSameState(const EntityTransformState* other) const
{
    const LightTransformState* state =
        dynamic_cast<const LightTransformState*>(other);
    if ( state == nullptr ) {
        return false;
    }
    return state->light == light && position.equals(state->position);
}

EntityTransformState* LightTransformState::clone() const
{
    return new LightTransformState(*this);
}
